proposals/0004-FaultTolerance.md

```diff
@@ -0,0 +1,224 @@
+
+By default, a failure handling strategy could assume, for example, that any exception is a failure. But in some cases, a `null` or negative return value could also be considered a failure. Users should be able to define this, and a user's definition of a failure is what should influence execution. (This all is what the Failsafe RetryPolicy's `retryOn`, `retryWhen`, `abortIf`, etc methods are all about - defining a failure).
+
+Standardise the Fallback, Bulkhead and CircuitBreaker APIs and provide implementations.
+
+* A Java API to provide RetryPolicy
+* A Java API to provide FallBack
+* A Java API to provide BulkHead
+* A Java API to provide CircuitBreaker
+
+## Detailed design
+
+
+### CDI-based approach 
+Use interceptor and annotation to specify the execution and policy configration.
+Define an interceptor binding to specify whether the execution is sync or asyn. Two interceptors should be provided by the implementation: AsyncExecutorInterceptor and SyncExecutorInterceptor
+```
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.METHOD, ElementType.TYPE })
+@InterceptorBinding
+@Inherited
+public @interface Execution {
+
+    Type value() default Type.SYNC;
+
+    enum Type {
+        ASYNC, SYNC;
+    };
+}
+```
+#### RetryPolicy: A policy to define the retry criteria
+
+An annotation to specify the max retries, delays, maxDuration, Duration unit, jitter, retryOn, bakeOff, fallback etc.
+```
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.METHOD, ElementType.TYPE })
+public @interface Retry {
+   
+    int maxRetries() default 3;
+
+    int delay() default 0;
+
+    TimeUnit delayUnit() default TimeUnit.SECONDS;
+
+    int maxDuration() default 20;
+    TimeUnit durationUnit() default TimeUnit.SECONDS;
+    int jitter() default 2;
+
+    int jitterFactor() default 2;
+
+    TimeUnit jitterDelayUnit() default TimeUnit.SECONDS;
+
+    int backOff() default 2;
+
+    int backOffFactor() default 2;
+
+    TimeUnit bakeOffUnit() default TimeUnit.SECONDS;
+
+    Class<? extends Throwable>[] retryOn() default { RuntimeException.class };
+
+    Class<? extends Throwable>[] aboartOn() default { RuntimeException.class };
+	String fallback() default "";
+}
+
+
+```
+#### CircuitBreaker: a rule to define when to close the circuit
+
+```
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Target({ ElementType.TYPE, ElementType.METHOD })
+public @interface CircuitBreaker {
+
+    Class<? extends Throwable>[] failOn() default RuntimeException.class;
+
+    long delay() default 2;
+
+    TimeUnit delayUnit() default TimeUnit.SECONDS;
+
+    long failThreshold() default 2;
+
+    long successThreshold() default 2;
+
+    long timeOut() default 2;
+
+}
+
+
+
+```
+#### Bulkhead - threadpool or semaphore styel
```

Comment:
After talking with Emily, I would be extremely concerned about bringing in Bulkheads in the first pass.  First, it implies some relationship between components.  Second, there is no way in CDI to do context propagation using built in contexts so the programming model becomes very weird.  It would almost certainly have to be blocking.